week6/isSubtree.cc

```diff
@@ -0,0 +1,160 @@
+/* isSubtree.cc
+ * author: connie
+ *
+ * IS SUBTREE
+ * Checks if a binary tree T is a subtree of binary tree S.
+ * A subtree is defined as have the same structure as another tree's root
+ * or child nodes -
+ * i.e., T does not necessarily have to be a node actually in tree S.
+ *
+ */
+
+#include <cstdlib>
+#include <iostream>
+#include <cstdio>
+#include <cmath>
+#include <cassert>
+
+using namespace std;
+
+//== DEFINITIONS
+struct tNode // tNode = binary treeNode
+{
+  int data;
+  tNode *left;
+  tNode *right;
+  tNode(int d) : data(d), left(NULL), right(NULL) {}
+};
+
+//== DECLARATIONS
+void deleteTree(tNode*& root);
+bool isSameTree(tNode *t1, tNode *t2);
+bool isSubtree(tNode *s, tNode *t);
+
+//== FUNCTIONS
+/*
+ * void deleteTree(tNode*& root)
+ * inputs: root - root node of the tree that we want to delete
+ * output: none
+ * notes: post-order traversal helper. don't need those memory leaks :P
+ */
+void deleteTree(tNode*& root)
+{
+  if (root) {
+    deleteTree(root->left);
+    deleteTree(root->right);
+    delete root;
+    root = NULL;
+  }
+}
+
+/*
+ * bool isSameTree(tNode *t1, tNode *t2)
+ * inputs: *t - pointer to the root of first tree
+           *s - pointer to the root of second tree
+ * output: true if t1 and t2 are structurally the same, false otherwise
+ * notes: helper function to validate tree structure
+ */
+bool isSameTree(tNode *t1, tNode *t2)
+{
+  // base case: both null
+  if (!t1 && !t2) return true;
+
+  else {
+    // if either node is null, then t1 and t2 are structurally different
+    if (!t1 || !t2) return false;
+    // next validate that the contents of the node are the same
+    if (t1->data == t2->data) {
+      // only recurse if we get this far
+      return isSameTree(t1->left, t2->left) && isSameTree(t1->right, t2->right);
+    }
+  }
+
+  // shouldn't ever get this far, but in case we do, return false to be safe
+  return false;
+}
+
+/*
+ * bool isSubtree(tNode *s, tNode *t)
+ * inputs: *t - pointer to the root of the subtree we are validating
+           *s - pointer to the root of the tree we are using as reference
+ * output: true if t is a subtree of s, false otherwise
+ * notes: main solution function
+ */
+bool isSubtree(tNode *t, tNode *s)
+{
+  // base case + null cases
+  if (!s && !t) return true;
+  else if (!s || !t) return false;
+
+  else {
+    // only check isSameTree if the node data is the same
+    if ((s->data == t->data) && isSameTree(s,t)) return true;
+
+    // otherwise:
+    // recurse on L and R children of s to find another suitable node
+    // as long as either side comes up with a matching subtree, we are done
+    else return (isSubtree(t, s->left) || isSubtree(t, s->right));
+  }
+}
```

Comment:
Great job on the implementation! Your work is always extremely readable (well-structured and nicely commented) which I very much appreciate :)